src/Microsoft.VisualStudio.Services.Agent/AgentServer.cs

```diff
@@ -41,44 +54,120 @@ public interface IAgentServer : IAgentService
+        private VssConnection _messageConnection;
+        private VssConnection _requestConnection;
+        private TaskAgentHttpClient _genericTaskAgentClient;
+        private TaskAgentHttpClient _messageTaskAgentClient;
+        private TaskAgentHttpClient _requestTaskAgentClient;
+
+        public async Task ConnectAsync(Uri serverUrl, VssCredentials credentials)
         {
             if (HostContext.RunMode == RunMode.Local)
             {
                 return;
             }
 
-            _connection = agentConnection;
+            _genericConnection = await EstablishVssConnection(serverUrl, credentials, TimeSpan.FromSeconds(100));
+            _messageConnection = await EstablishVssConnection(serverUrl, credentials, TimeSpan.FromSeconds(60));
+            _requestConnection = await EstablishVssConnection(serverUrl, credentials, TimeSpan.FromSeconds(60));
+
+            _genericTaskAgentClient = _genericConnection.GetClient<TaskAgentHttpClient>();
+            _messageTaskAgentClient = _messageConnection.GetClient<TaskAgentHttpClient>();
+            _requestTaskAgentClient = _requestConnection.GetClient<TaskAgentHttpClient>();
+
+            _hasGenericConnection = true;
+            _hasMessageConnection = true;
+            _hasRequestConnection = true;
+        }
+
+        public async Task RefreshConnectionAsync(AgentConnectionType connectionType, TimeSpan timeout)
+        {
+            Trace.Info($"Refresh {connectionType} VssConnection to get on a different AFD node.");
+            switch (connectionType)
+            {
+                case AgentConnectionType.MessageQueue:
+                    _hasMessageConnection = false;
+                    _messageConnection = await EstablishVssConnection(_messageConnection.Uri, _messageConnection.Credentials, timeout);
+                    _messageTaskAgentClient = _messageConnection.GetClient<TaskAgentHttpClient>();
+                    _hasMessageConnection = true;
+                    break;
+                case AgentConnectionType.JobRequest:
+                    _hasRequestConnection = false;
+                    _requestConnection = await EstablishVssConnection(_requestConnection.Uri, _requestConnection.Credentials, timeout);
+                    _requestTaskAgentClient = _requestConnection.GetClient<TaskAgentHttpClient>();
+                    _hasRequestConnection = true;
+                    break;
+                default:
+                    throw new NotSupportedException(connectionType.ToString());
+            }
+        }
+
+        public void SetConnectionTimeout(AgentConnectionType connectionType, TimeSpan timeout)
+        {
+            Trace.Info($"Set {connectionType} VssConnection's timeout to {timeout.TotalSeconds} seconds.");
+            switch (connectionType)
+            {
+                case AgentConnectionType.JobRequest:
+                    _requestConnection.Settings.SendTimeout = timeout;
+                    break;
+                default:
```

Comment:
it may be better to just implement the other two cases since this method is only called during exception flow